0010-zkvm-keccak.md

```diff
@@ -0,0 +1,788 @@
+
+#### Negation
+
+The NOT operation will be performed with a subtraction operation using the constant term $(2^{16}-1)$ for each 16-bit quarter, and only later it will be expanded. The mechanism to check that the input is at most 64 bits long, will rely on the composition of the sparse representation. Meaning, 64-bit words produce four 16-bit lookups, which implies that the real bits take no more than 64 bits. Given $x$ in reset form (meaning, not any sparse representation of $x$ but the initial expanded one with zero intermediate bits) the negation of one 64-bit word can be constrained as:
+
+$$\forall i \in[0, 3]: expand(not_i) = 0\text{x}1111111111111111 - expand(x_i)$$
+
+> Because the inputs are known to be 16-bits at most, it is guaranteed that the total length of $x$ is at most 64 bits, and is correctly split into quarters `x[i]`.
+
+#### Rotation
+
+Rotations of $b$ bits to the left will be performed directly on the full 64-bit word $X$ to obtain $Y$, using the algebraic observation that:
+
+$$X \cdot 2^b = Q \cdot 2^{64} + R \quad , \quad Y = Q + R$$ 
+
+Making use of an auxiliary function to compose four 16-bit quarters into full 64-bit terms as:
+
+$$compose \to quarter_0 + 2^{16}quarter_1 + 2^{32}quarter_2 + 2^{48}quarter_3$$
+
+Then, the rotation of `offset` bits to the left can be constrained as:
+
+```rust
+constrain(0 == input * two_to_off - (quotient * 2^64 + remainder))
+constrain(0 == remainder + quotient - output)
+
+// Check quotient < 2^offset <=> quotient + 2^64 - 2^off < 2^64
+constrain(0 == bound - ( quotient + 2^64 - two_to_off ) )
+```
+
+Together with $3\times4=12$ lookups to check that the remainder, quotient and bound are $<2^{64}$.
+
+> Assuming that we execute rotation starting form reset values, then one could perform the above operations directly on the expanded terms, since $4\cdot64-3=253$ encoded bits (removing the three leading zeros) would fit in the BN254 curve. In this scenario, rotations by $x$ bits would be mapped to rotations by an offset of $4\cdot x$ bits instead. Moreover, the constraints would need to be slightly different, to avoid overflows. In particular: `word = quotient * 2^(253-4*offset) + shift` and `output = quotient + shift * 2^(4*offset)`, and the output would already be in expanded format.  Nonetheless, in this case we would also need to perform bound checks, which would be much more costly due to the bitlengths. And in any case, each dense word should be expanded afterwards to be processed in later steps.
+
+### Constraints
+
+
+> For efficiency reasons, the state is assumed to be stored in expanded form, so that back-and-forth conversions do not need to take place repeatedly.
+
+The support for the Keccak hash function will require the following gate types:
+
+**PRECONDITIONS:** the message to be hashed should be padded with the $10^*1$ rule before 
```

Comment:
Is it possible to add a sentence (like in Appendix or redirect to a place describe it) explaining/formalizing the padding rule?